thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/dashboard/resources/AnomalyFlattenResource.java

```diff
@@ -23,87 +23,205 @@
+      MetricSlice metricSlice = MetricSlice.from(metricId, start, end);
+      DataFrame resultDataFrame = null;
+      try {
+        resultDataFrame = aggregationLoader.loadAggregate(metricSlice, dimensionKeys, -1);
+      } catch (Exception e) {
+        throw new ExecutionException(e);
+      }
+      return resultDataFrame;
+    });
   }
 
   /**
-   * Flat an anomaly to a flat map structure
-   * @param anomalyResult an instance of MergedAnomalyResultDTO
-   * @param tableKeys a list of keys in dimensions; if null, use all keys in dimension
-   * @return a map of information in the anomaly result with the required keys
+   * Reformat the rows in data frame and anomaly information into a list of map
+   * @param metricDataFrame a map from metric name to dataframe with aggregated time series
+   * @param anomalies a list of anomalies within the same time interval as dataframe
+   * @param dimensions a list of dimensions
+   * @return a list of maps with dataframe and anomaly information
    */
-  public static Map<String, Object> flatAnomalyResult(MergedAnomalyResultDTO anomalyResult,
-      List<String> tableKeys) {
-    Preconditions.checkNotNull(anomalyResult);
-    Map<String, Object> flatMap = new HashMap<>();
-    flatMap.put(ANOMALY_ID, anomalyResult.getId());
-    DimensionMap dimension = anomalyResult.getDimensions();
-    if (tableKeys == null) {
-      tableKeys = new ArrayList<>(dimension.keySet());
+  public static List<Map<String, Object>> reformatDataFrameAndAnomalies(Map<String, DataFrame> metricDataFrame,
+      List<MergedAnomalyResultDTO> anomalies, List<String> dimensions) {
+    List<Map<String, Object>> resultList = new ArrayList<>();
+    Map<DimensionMap, Map<Long, String>> anomalyCommentMap = extractComments(anomalies);
+    List<String> metrics = new ArrayList<>(metricDataFrame.keySet());
+    Map<DimensionMap, Map<String, Double>> metricValues = new HashMap<>();
+    for (String metric : metrics) {
+      DataFrame dataFrame = metricDataFrame.get(metric);
+      for (int i = 0; i < dataFrame.size(); i++) {
+        DimensionMap dimensionMap = new DimensionMap();
+        for (String dimensionKey : dimensions) {
+          String dimensionValue = dataFrame.getString(dimensionKey, i);
+          dimensionMap.put(dimensionKey, dimensionValue);
+        }
+        if (!metricValues.containsKey(dimensionMap)) {
+          metricValues.put(dimensionMap, new HashMap<>());
+        }
+        metricValues.get(dimensionMap).put(metric, dataFrame.getDouble(DATAFRAME_VALUE, i));
+      }
     }
-    for (String key : tableKeys) {
-      flatMap.put(key, dimension.getOrDefault(key, ""));
+
+    for (DimensionMap dimensionMap : metricValues.keySet()) {
+      Map<String, Object> resultMap = new HashMap<>();
+      for (String key : dimensionMap.keySet()) {
+        resultMap.put(key, dimensionMap.get(key));
+      }
+      for (String metric : metrics) {
+        if (metricValues.containsKey(dimensionMap)) {
```

Comment:
remove the if/else here. 
just ```resultMap.put(metric, metricValues.get(dimensionMap).getOrDefault(metric, Double.NaN));```  is good enough.

Reply:
If dimensionMap is not in metricValues, metricValues.get(dimensionMap) will return null. NPE will be thrown with the process.